artiq/firmware/ksupport/eh_artiq.rs

```diff
@@ -352,3 +361,29 @@ pub fn get_exception_id(name: &str) -> u32 {
     unimplemented!("unallocated internal exception id")
 }
 
+/// Takes as input exception id from host
+/// Generates a new exception with:
+///   * `id` set to `exn_id`
+///   * `message` set to corresponding exception name from `EXCEPTION_ID_LOOKUP`
+///
+/// The message is matched on host to ensure correct exception is being referred 
+/// This test checks the synchronization of exception ids for runtime errors
+#[no_mangle]
+pub extern "C-unwind" fn test_exception_id_sync(exn_id: u32) {
+    let message = EXCEPTION_ID_LOOKUP
+        .iter()
+        .find_map(|&(name, id)| if id == exn_id { Some(name) } else { None })
+        .unwrap_or("unallocated internal exception id");
```

Comment:
Raising an exception outside the provided list already crashes the kerenel: https://github.com/m-labs/artiq/blob/990df7fa775c2df7cbacc2779cb7869365ad5965/artiq/firmware/ksupport/eh_artiq.rs#L361
Since, this part is used for testing the id sync, it would fail those tests but otherwise will not crash the kernel. Also, raising errors on firmware requires explicitly calling `artiq_raise` with the specified exception name and so there are no random errors due to user inputs as they would all be generated on the host side.

Reply:
Would it still be possible to crash the kernel CPU if an ID is added to one side but not the other? The concern would be while testing new changes.

Reply:
> Would it still be possible to crash the kernel CPU if an ID is added to one side but not the other? The concern would be while testing new changes; the user wouldn't have to call this function at all anyway.

While this could be done for the ID part, the main use of the new syscall is to match exception names to Ids. So, for instance exception Id 0 on both kernel and host should point to `RTIOUnderflow`. The function `get_exception_id`:
https://github.com/m-labs/artiq/blob/990df7fa775c2df7cbacc2779cb7869365ad5965/artiq/firmware/ksupport/eh_artiq.rs#L361
would crash the firmware if at runtime an exception outside the list is raised.